stew/result.nim

```diff
@@ -232,20 +237,25 @@ template `and`*(self: Result, other: untyped): untyped =
 template `or`*(self: Result, other: untyped): untyped =
   ## Evaluate `other` iff not self.isOk, else return self
   ## fail-fast - will not evaluate other if a is a value
+  ##
+  ## TODO: This API is unsafe due to potential multiple
+  ## evaluation of the `self` parameter.
   if self.isOk: self
   else: other
 
-template catch*(body: typed): Result[type(body), ref Exception] =
+template catch*(body: typed): Result[type(body), ref CatchableError] =
```

Comment:
Sync `catch` is a synchronous mechanism, I think it shouldn't stop the propagation of non-recoverable errors by default. There might be another API called `catchAllExceptions` that captures defects as well.

Reply:
this code is pre-`CatchableError`, so this is indeed a good update - the question perhaps is what to do with the rest of the raw `Exception` support in the library - one way could be to have special "wrapping" or migration code for stuff that inherits from `Exception` but not `Catchable`. same thing with `when E is ref Exception` above - should it limit itself to CatchableError? if yes, the exception list of `raiseResultError` could be bounded better.


I'd generally consider all use of `Defect` UB in terms of behaviour and implementation (ie see manual and the latest `panic` patch) thus it seems wrong to cement "support" for it here by adding `catchAllExceptions` - or if we do add it, perhaps it should be `deprectated` from the start to mark it a support for dealing with legacy code more than anything else.